### PR TITLE
Ignore non-draw codes when calculating path extent

### PR DESCRIPTION
**Pull request by dstansby**
When calculating the extents of a path, any ~MOVETO,~ CLOESPOLY, or STOP codes shouldn't be taken into account.  This is a follow up to https://github.com/matplotlib/matplotlib/pull/16832.

### CODE REVIEW

**Comment by dstansby** [comment]
🤔 looking at the failing tests it looks like any MOVETO codes are deliberately included in the path extents. This should probably just ignore CLOSEPOLY and STOP.

File: lib/matplotlib/path.py [hunk 1]
```diff
@@ -620,11 +623,20 @@ def get_extents(self, transform=None, **kwargs):
...
+        # Short cut the case with only straight lines
         elif len(np.intersect1d(self.codes, [Path.CURVE3, Path.CURVE4])) == 0:
-            xys = self.vertices[self.codes != Path.CLOSEPOLY]
+            draw_idxs = np.in1d(self.codes, Path.LINETO)
+            # Include the start and end point of each line
+            draw_idxs = draw_idxs | np.roll(draw_idxs, -1)
+            # Add in commands that aren't ignored
+            # (in the straight line case here this is only MOVETO)
+            draw_idxs = draw_idxs | ~np.in1d(self.codes, Path.IGNORED_CODES)
+            xys = self.vertices[draw_idxs]
```
Comment (timhoffm) [inline lib/matplotlib/path.py:634]: IMHO this way of writing is more concise:

```suggestion
            # we have only straight lines:
            # consider line starts, line ends and points moved to
            line_end_mask = np.isin(self.codes, Path.LINETO)
            line_start_mask = np.roll(line_end_mask, -1)
            moveto_mask = np.isin(self.codes, Path.MOVETO)
            xys = self.vertices[line_start_mask | line_end_mask | moveto_mask]
```

Note also that I've used `np.isin()` instead of `np.in1d()` because `self.codes` are 1D.

*Edit*: I'm wondering if 
~~~
xys = self.vertices[np.isin(self.codes, [Path.LINETO, Path.MOVETO])]
~~~
would be sufficient. Here, the codes of points at line starts cannot be STOP, CURVE3, CURVE4. The only grey area would be CLOSEPOLY followed by LINETO, in which case the line start would be a CLOSEPOLY node. IMHO that's not quite well defined CLOSEPOLY should always be followed by MOVETO.

File: lib/matplotlib/path.py [hunk 2]
```diff
@@ -620,11 +623,20 @@ def get_extents(self, transform=None, **kwargs):
             self = transform.transform_path(self)
         if self.codes is None:
             xys = self.vertices
+        # Short cut the case with only straight lines
```
Comment (timhoffm) [inline lib/matplotlib/path.py:626]: Should be moved into the elif block. See the above comment for a possible wording.

File: lib/matplotlib/path.py [hunk 3]
```diff
@@ -95,6 +95,9 @@ class Path:
                              CURVE4: 3,
                              CLOSEPOLY: 1}
 
+    # A list of path codes for which the vertices are ignored
+    IGNORED_CODES = [STOP, CLOSEPOLY]
```
Comment (timhoffm) [inline lib/matplotlib/path.py:99]: Maybe
```suggestion
    CODES_WITH_IGNORED_VERTICES = [STOP, CLOSEPOLY]
```
A bit lengthy but the name is clearer.

File: lib/matplotlib/path.py (hunk 1 above)
Comment (dstansby) [reply lib/matplotlib/path.py]: I think they can be CURVE{3,4}, as there will always be a final control point for those curves (the curve endpoint) which one can then draw a straight line from.

**Comment by dstansby** [comment]
👍 thanks for the suggestions, agreed that they make the code read much better

File: lib/matplotlib/path.py (hunk 1 above)
Comment (timhoffm) [reply lib/matplotlib/path.py]: But the `elif` conditions explicitly excludes CURVE{3,4} here.

Comment (brunobeltran) [reply lib/matplotlib/path.py]: Sorry for the lag. A couple of things: 

> The only grey area would be CLOSEPOLY followed by LINETO, in which case the line start would be a CLOSEPOLY node. IMHO that's not quite well defined CLOSEPOLY should always be followed by MOVETO.

While "undefined" is a reasonable guess, AFAICT, all backends currently handle this behavior consistently. That is, CLOSEPOLY's vertex is ignored and it is treated as "(start_point, LINETO)" plus adding the extra cap. 

Which means that CLOSEPOLY followed by LINETO is well-defined, and does the natural thing:
```python
import matplotlib.pyplot as plt
from matplotlib.path import Path
from matplotlib.patches import PathPatch

p = Path(
    [[0, 0], [0, 1], [1, 1], [5, 5], [1, 0], [0, -1], [-5, 5]], 
    [Path.MOVETO, Path.LINETO, Path.LINETO, Path.CLOSEPOLY, 
     Path.LINETO, Path.LINETO, Path.CLOSEPOLY]
)
fig, ax = plt.subplots()
ax.add_patch(PathPatch(p))
ax.set_xlim([-0.5, 1.5])
ax.set_ylim([-1.5, 1.5])
```
image

And, unless I'm missing something, then 


> I'm wondering if
> 
> ```
> xys = self.vertices[np.isin(self.codes, [Path.LINETO, Path.MOVETO])]
> ```
> 
> would be sufficient. 

tl;dr: yes. We should just be doing this.

File: lib/matplotlib/tests/test_path.py [hunk 4]
```diff
@@ -102,6 +102,15 @@ def test_exact_extents(path, extents):
     assert np.all(path.get_extents().extents == extents)
 
 
+@pytest.mark.parametrize('ignored_code', [Path.CLOSEPOLY, Path.STOP])
+def test_extents_with_ignored_codes(ignored_code):
+    # Check that STOP and CLOSEPOLY points are ignored when caclulating extents
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_path.py:107]:
```suggestion
    # Check that STOP and CLOSEPOLY points are ignored when calculating extents
```

File: lib/matplotlib/path.py [hunk 5]
```diff
@@ -95,6 +95,9 @@ class Path:
                              CURVE4: 3,
                              CLOSEPOLY: 1}
 
+    # A list of path codes for which the vertices are ignored
+    CODES_WITH_IGNORED_VERTICES = [STOP, CLOSEPOLY]
```
Comment (QuLogic) [inline lib/matplotlib/path.py:99]: Does this need to be public? Then it should probably have a docstring (`#:`).

**Review by QuLogic — COMMENTED** [review]
Should the ignored code list be integrated into #19088?

File: lib/matplotlib/path.py [hunk 6]
```diff
@@ -621,10 +624,17 @@ def get_extents(self, transform=None, **kwargs):
...
+            # Only need to consider line starts, line ends and points moved to
+            line_end_mask = np.isin(self.codes, Path.LINETO)
+            line_start_mask = np.roll(line_end_mask, -1)
+            moveto_mask = np.isin(self.codes, Path.MOVETO)
+            xys = self.vertices[line_start_mask | line_end_mask | moveto_mask]
         else:
             xys = []
             for curve, code in self.iter_bezier(**kwargs):
+                if code in Path.CODES_WITH_IGNORED_VERTICES:
+                    continue
```
[the remark this answers is not in the document]
Comment (brunobeltran) [reply lib/matplotlib/path.py:637]: tl;dr: if you reverted this change the code would still be "correct", AFAICT.

**Review by brunobeltran — CHANGES_REQUESTED** [review]
Thanks for this catch @dstansby! 

Not sure why this wasn't just `xys = self.vertices[np.isin(self.codes, [Path.MOVETO, Path.LINETO])]` right from the start.

File: lib/matplotlib/path.py (hunk 5 above)
Comment (dstansby) [reply lib/matplotlib/path.py:99]: It doesn't need to be, but it might supplement the main docstring above to have this public and documented.

Comment (dstansby) [reply lib/matplotlib/path.py:99]: Ah, I've removed it as per the comment below

**Comment by dstansby** [comment]
Thanks for the reviews. There were enough changes that I've just ammended the commit and force pushed a new version.

**Review by timhoffm — APPROVED** [review]

**Review by brunobeltran — APPROVED** [review]
LGTM.
